Python/pystate.c

```diff
@@ -293,7 +294,8 @@ _PyInterpreterState_DeleteExceptMain()
         if (interp->id_mutex != NULL) {
             PyThread_free_lock(interp->id_mutex);
         }
-        PyMem_RawFree(interp);
+        tmp_interp = interp;
+        PyMem_RawFree(tmp_interp);
```

Comment:
I don't understand how the current code can work since PyMem_RawFree() fills freed memory with a byte pattern 0xDB and so interp = interp->next should just crash.

I don't see how this change fix anything: interp still points to freed bytes.

You should read interp->next before PyMem_RawFree(), and so change how the loop is written.

cc @ericsnowcurrently

Reply:
When  ```PyMem_RawFree(tmp_interp)``` the ```tmp_interp```  point to ```interp```, so you are freed ```interp``` too, right?

What about this?
```C
tmp_interp = interp->next;
for (; tmp_interp !=NULL; interp = tmp_interp){
...
tmp_interp = interp->next;
PyMem_RawFree(interp);
}

```

Reply:
sure? the last time I have coded in C, that was in 2007 :/ ok

Reply:
or just
```c
PyInterpreterState *tmp_interp = NULL;
for (; interp != NULL;) {
    tmp_interp = interp;
    interp = interp->next;
    PyMem_RawFree(tmp_interp);
}
```

Reply:
@eamanu are you on Zulip or IRC?

Reply:
@matrixise I am connecting to zulip

Reply:
With a `while (interp)` should work the same, and the jungling of tmp/interp looks good to me.

Reply:
@JulienPalard I have just updated my PR, thanks for the review.
Now, I am going to wait for the review of @vstinner, he is more expert than me with the C language.